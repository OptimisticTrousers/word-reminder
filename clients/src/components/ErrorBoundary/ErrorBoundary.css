.boundary {
  padding: 1rem;
}

.boundary__container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.boundary__image {
  width: 250px;
  height: 250px;
  margin-bottom: 1rem;
}

.boundary__title {
  font-size: 1.75rem;
  font-weight: 600;
  color: var(--heading-color);
}

.boundary__message {
  font-size: 1.25rem;
}

.boundary__title,
.boundary__message {
  margin-bottom: 0.75rem;
}
