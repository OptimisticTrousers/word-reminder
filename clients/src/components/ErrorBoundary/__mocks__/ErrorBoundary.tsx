import { Component, ReactNode } from "react";

interface Props {
  children: ReactNode;
}

interface State {
  error: Error | null;
}

export class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      error: null,
    };
  }

  static getDerivedStateFromError(error: Error) {
    return { error };
  }

  render() {
    if (this.state.error) {
      return <div data-testid="error-boundary"></div>;
    }
    return this.props.children;
  }
}
