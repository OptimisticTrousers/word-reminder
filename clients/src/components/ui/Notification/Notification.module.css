.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --text-color: var(--success-text-color);
  --icon-color: var(--success-icon-color);
  --button-bkg-color: var(--success-button-bkg-color);
  --secondary-button-bkg-color: var(--secondary-success-button-bkg-color);
  background-color: var(--success-bkg-color);
}

.notification--error {
  --text-color: var(--alert-text-color);
  --icon-color: var(--alert-icon-color);
  --button-bkg-color: var(--alert-button-bkg-color);
  --secondary-button-bkg-color: var(--secondary-alert-button-bkg-color);
  background-color: var(--alert-bkg-color);
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--text-color);
  padding: 0.375rem;
  background-color: var(--button-bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__button:hover {
  background-color: var(--secondary-button-bkg-color);
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
