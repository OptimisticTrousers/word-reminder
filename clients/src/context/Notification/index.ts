export { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
export { NotificationProvider } from "./Provider";
