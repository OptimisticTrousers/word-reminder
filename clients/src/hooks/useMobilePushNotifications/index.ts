export { useMobilePushNotificationRegistration } from "./useMobilePushNotifications";
export { useMobilePushNotificationListeners } from "./useMobilePushNotifications";
