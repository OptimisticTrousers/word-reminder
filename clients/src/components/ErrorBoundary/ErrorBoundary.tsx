import { Component, ReactNode } from "react";
import { Link } from "react-router-dom";

import "./ErrorBoundary.css";

interface Props {
  children: ReactNode;
}

interface State {
  error: Error | null;
}

export class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      error: null,
    };
  }

  static getDerivedStateFromError(error: Error) {
    return { error };
  }

  render() {
    if (this.state.error) {
      return (
        <div className="boundary">
          <div className="boundary__container">
            <img
              className="boundary__image"
              src="/images/word-reminder.png"
              alt=""
            />
            <h2 className="boundary__title">An error occurred.</h2>
            <p className="boundary__message">{this.state.error.message}</p>
            <Link to="/" className="boundary__link">
              Go back to the home page.
            </Link>
          </div>
        </div>
      );
    }
    return this.props.children;
  }
}
