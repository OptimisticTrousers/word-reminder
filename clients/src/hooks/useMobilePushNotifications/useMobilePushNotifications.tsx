import {
  PushNotifications,
  Token,
  ActionPerformed,
  RegistrationError,
} from "@capacitor/push-notifications";
import { useContext, useEffect } from "react";
import { NotificationContext } from "../../context/Notification";
import { NavigateFunction, useNavigate } from "react-router-dom";
import { fcmTokenService } from "../../services/fcm_token_service";
import { Capacitor } from "@capacitor/core";

const addListeners = async (
  navigate: NavigateFunction,
  showNotification: (type: string, message: string) => void,
  userId: string
) => {
  // On success, we should be able to receive notifications
  await PushNotifications.addListener("registration", (token: Token) => {
    fcmTokenService.createFCMToken({ token, userId });
  });

  // Some issue with our setup and push will not work
  await PushNotifications.addListener(
    "registrationError",
    (error: RegistrationError) => {
      showNotification("error", error.error);
    }
  );

  // Method called when tapping on a notification
  await PushNotifications.addListener(
    "pushNotificationActionPerformed",
    (notification: ActionPerformed) => {
      const wordReminderId = notification.notification.data.wordReminderId;
      navigate(`/wordReminders/${wordReminderId}`);
    }
  );
};

export function useMobilePushNotificationRegistration(userId: string) {
  const { showNotification } = useContext(NotificationContext);
  const navigate = useNavigate();

  async function register() {
    if (Capacitor.getPlatform() === "web") {
      return;
    }
    // Register with Apple / Google to receive push via APNS/FCM
    const status = await PushNotifications.checkPermissions();

    if (status.receive !== "granted") {
      const requestStatus = await PushNotifications.requestPermissions();
      if (requestStatus.receive === "denied") {
        showNotification("error", "Push Notification permission denied.");
      } else {
        // Register with Apple / Google to receive push via APNS/FCM
        await PushNotifications.register();
        await addListeners(navigate, showNotification, userId);
      }
    }
  }

  return { register };
}

export function useMobilePushNotificationListeners(userId: string) {
  const { showNotification } = useContext(NotificationContext);
  const navigate = useNavigate();

  useEffect(() => {
    (async () => {
      if (Capacitor.getPlatform() === "web") {
        return;
      }
      const status = await PushNotifications.checkPermissions();
      if (status.receive === "granted") {
        // Register with Apple / Google to receive push via APNS/FCM
        await PushNotifications.register();
        await addListeners(navigate, showNotification, userId);
      }
    })();

    return () => {
      if (Capacitor.getPlatform() === "web") {
        return;
      }
      PushNotifications.removeAllListeners();
    };
  }, [navigate, showNotification, userId]);
}
