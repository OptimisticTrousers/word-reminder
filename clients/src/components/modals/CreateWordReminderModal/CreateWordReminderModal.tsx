import { Detail, User, UserWord } from "common";
import CSSModules from "react-css-modules";
import { useOutletContext } from "react-router-dom";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { ModalContainer } from "../ModalContainer";
import { userWordService } from "../../../services/user_word_service";
import styles from "./CreateWordReminderModal.module.css";
import { ReactNode, useContext, useMemo } from "react";
import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import { useNotificationError } from "../../../hooks/useNotificationError";
import { wordReminderService } from "../../../services/word_reminder_service";
import { ToggleModal } from "../types";
import { Reminder } from "../../ui/Reminder";
import { useMobilePushNotificationRegistration } from "../../../hooks/useMobilePushNotifications";

interface Props {
  toggleModal: ToggleModal;
}

export const CreateWordReminderModal = CSSModules(
  function ({ toggleModal }: Props) {
    const { user }: { user: User } = useOutletContext();
    const userId = String(user.id);
    const { showNotification } = useContext(NotificationContext);
    const { showNotificationError } = useNotificationError();
    const { register } = useMobilePushNotificationRegistration(userId);
    const queryClient = useQueryClient();
    const { data } = useQuery({
      queryKey: ["userWords"],
      queryFn: async () => {
        return userWordService.getUserWordList({
          userId,
          params: Object.fromEntries(new URLSearchParams()),
        });
      },
      throwOnError: true,
    });
    const { isPending, mutate } = useMutation({
      mutationFn: wordReminderService.createWordReminder,
      onSuccess: () => {
        showNotification(
          NOTIFICATION_ACTIONS.SUCCESS,
          CREATE_WORD_REMINDER_NOTIFICATION_MSGS.createWordReminder()
        );
        queryClient.invalidateQueries({
          queryKey: ["wordReminders"],
        });
        register();
      },
      onError: showNotificationError,
      onSettled: () => {
        toggleModal();
      },
    });

    const [wordToUserWord, words] = useMemo(() => {
      const wordToUserWord: { [key: string]: UserWord } = {};

      const words: ReactNode[] = data?.json.userWords.map(
        (userWord: UserWord & { details: Detail[] }) => {
          const word = userWord.details[0].word;
          wordToUserWord[word] = userWord;
          return (
            <option key={userWord.id} styleName="modal__option" value={word}>
              {word}
            </option>
          );
        }
      );
      return [wordToUserWord, words];
    }, [data]);

    function handleCreate(formData: FormData) {
      const userWords = formData.get("user_words") as string;
      const userWordsData = userWords.split(",").map((userWord: string) => {
        return wordToUserWord[userWord.toLowerCase()];
      });

      mutate({
        userId,
        body: {
          reminder: formData.get("reminder") as string,
          finish: new Date(formData.get("finish") as string),
          is_active: (formData.get("is_active") as string) === "on",
          has_reminder_onload:
            (formData.get("has_reminder_onload") as string) === "on",
          user_words: userWordsData,
        },
      });
    }

    return (
      <ModalContainer title="Create Word Reminder" toggleModal={toggleModal}>
        <form styleName="modal" action={handleCreate}>
          <Reminder disabled={false} value={""} />
          <div styleName="modal__control">
            <label styleName="modal__label" htmlFor="finish">
              Finish
            </label>
            <input
              styleName="modal__input"
              type="date"
              id="finish"
              name="finish"
              required
            />
          </div>
          <div styleName="modal__control">
            <label styleName="modal__label" htmlFor="user_words">
              User Words
            </label>
            <input
              styleName="modal__input"
              type="email"
              multiple
              name="user_words"
              id="user_words"
              autoFocus
              onFocus={(event) => {
                event.target.type = "email";
              }}
              onBlur={(event) => {
                event.target.type = "text";
              }}
              list="words"
              required
              size={64}
              placeholder="Separate words with a comma and no spaces."
            />
            <datalist id="words">{words}</datalist>
          </div>
          <fieldset styleName="modal__fieldset">
            <legend styleName="modal__legend">Options</legend>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="is_active">
                Is Active
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="is_active"
                name="is_active"
                defaultChecked
              />
            </div>
            <div styleName="modal__control">
              <label styleName="modal__label" htmlFor="has_reminder_onload">
                Has Reminder Onload
              </label>
              <input
                styleName="modal__input"
                type="checkbox"
                id="has_reminder_onload"
                name="has_reminder_onload"
                defaultChecked
              />
            </div>
          </fieldset>
          <button styleName="modal__button" disabled={isPending} type="submit">
            Create
          </button>
        </form>
      </ModalContainer>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);

const CREATE_WORD_REMINDER_NOTIFICATION_MSGS = {
  createWordReminder: () => {
    return "A word reminder has been created!";
  },
};
