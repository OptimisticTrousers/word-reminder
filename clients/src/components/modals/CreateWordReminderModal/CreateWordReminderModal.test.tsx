import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createRoutesStub, Outlet } from "react-router-dom";
import { wordReminderService } from "../../../services/word_reminder_service";
import { NotificationProvider } from "../../../context/Notification";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { CreateWordReminderModal } from "./CreateWordReminderModal";
import { userWordService } from "../../../services/user_word_service";
import { ErrorBoundary } from "../../ErrorBoundary/ErrorBoundary";
import { Mock } from "vitest";
import * as useMobilePushNotificationHooks from "../../../hooks/useMobilePushNotifications";

vi.mock("../../ErrorBoundary/ErrorBoundary");
vi.mock("../ModalContainer/ModalContainer");

describe("CreateWordReminderModal component", () => {
  const testUser = {
    id: 1,
  };
  const word1 = {
    id: 1,
    details: [
      {
        word: "hello",
        phonetics: [
          {
            text: "hɛˈləʊ",
          },
        ],
        meanings: [
          {
            partOfSpeech: "exclamation",
            definitions: [
              {
                definition:
                  "used as a greeting or to begin a phone conversation.",
                example: "hello there, Katie!",
                synonyms: [],
                antonyms: [],
              },
            ],
          },
        ],
      },
    ],
    created_at: new Date(),
  };
  const word2 = {
    id: 2,
    details: [
      {
        word: "clemency",
        meanings: [
          {
            partOfSpeech: "noun",
            definitions: [{ definition: "Mercy; lenience." }],
          },
        ],
        phonetics: [],
      },
    ],
    created_at: new Date(),
  };

  const userWord1 = {
    id: 1,
    user_id: testUser.id,
    word_id: word1.id,
    details: word1.details,
    learned: false,
    created_at: new Date(),
    updated_at: new Date(),
  };

  const userWord2 = {
    id: 2,
    user_id: testUser.id,
    word_id: word2.id,
    details: word2.details,
    learned: false,
    created_at: new Date(),
    updated_at: new Date(),
  };

  const wordReminder = {
    id: 1,
    has_reminder_onload: false,
    is_active: false,
    user_id: testUser.id,
    reminder: "* * * * *",
    finish: new Date("2025-02-08"),
    user_words: [
      {
        learned: userWord1.learned,
        created_at: userWord1.created_at,
        updated_at: userWord1.updated_at,
        details: word1.details,
        id: userWord1.id,
      },
      {
        learned: userWord2.learned,
        created_at: userWord2.created_at,
        updated_at: userWord2.updated_at,
        details: word2.details,
        id: userWord2.id,
      },
    ],
    created_at: new Date(),
    updated_at: new Date(),
  };

  function setup({
    toggleModal,
    queryClient,
  }: {
    toggleModal: Mock;
    queryClient: QueryClient;
  }) {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <ErrorBoundary>
                  <NotificationProvider>
                    <QueryClientProvider client={queryClient}>
                      <CreateWordReminderModal toggleModal={toggleModal} />
                    </QueryClientProvider>
                  </NotificationProvider>
                </ErrorBoundary>
              );
            },
          },
        ],
      },
    ]);

    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={["/"]} />),
    };
  }

  const status = 200;

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders the form with all inputs", async () => {
    vi.spyOn(userWordService, "getUserWordList").mockImplementation(
      async () => {
        return { json: { userWords: [userWord1, userWord2] }, status };
      }
    );
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });

    const { asFragment } = setup({ toggleModal: mockToggleModal, queryClient });

    const reminder = screen.getByText("Reminder");
    const finish = screen.getByLabelText("Finish", { selector: "input" });
    const isActive = screen.getByLabelText("Is Active", { selector: "input" });
    const hasReminderOnload = screen.getByLabelText("Has Reminder Onload", {
      selector: "input",
    });
    const userWords = screen.getByRole("combobox");
    const userWordOption1 = await screen.findByText(userWord1.details[0].word);
    const userWordOption2 = screen.getByText(userWord2.details[0].word);
    const modalHeading = screen.getByTestId("modal-heading");
    expect(modalHeading).toBeInTheDocument();
    expect(modalHeading).toHaveTextContent("Create Word Reminder");
    expect(reminder).toBeInTheDocument();
    expect(finish).toBeInTheDocument();
    expect(isActive).toBeInTheDocument();
    expect(isActive).toBeChecked();
    expect(hasReminderOnload).toBeInTheDocument();
    expect(hasReminderOnload).toBeChecked();
    expect(userWords).toBeInTheDocument();
    expect(userWordOption1).toBeInTheDocument();
    expect(userWordOption1).toHaveValue(userWord1.details[0].word);
    expect(userWordOption2).toBeInTheDocument();
    expect(userWordOption2).toHaveValue(userWord2.details[0].word);
    expect(asFragment).toMatchSnapshot();
  });

  it("calls the functions to create a word reminder", async () => {
    vi.spyOn(userWordService, "getUserWordList").mockImplementation(
      async () => {
        return { json: { userWords: [userWord1, userWord2] }, status };
      }
    );
    const mockCreateWordReminder = vi
      .spyOn(wordReminderService, "createWordReminder")
      .mockImplementation(async () => {
        return { json: { wordReminder }, status };
      });
    const mockRegister = vi.fn();
    vi.spyOn(
      useMobilePushNotificationHooks,
      "useMobilePushNotificationRegistration"
    ).mockReturnValue({ register: mockRegister });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const reminder = screen.getByLabelText("Reminder");
    const finish = screen.getByLabelText("Finish", { selector: "input" });
    const isActive = screen.getByLabelText("Is Active", { selector: "input" });
    const hasReminderOnload = screen.getByLabelText("Has Reminder Onload", {
      selector: "input",
    });
    const userWords = await screen.findByRole("combobox");
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(reminder, "* * * * *");
    await user.type(finish, "2025-02-04");
    await user.type(
      userWords,
      `${userWord1.details[0].word.toUpperCase()},${userWord2.details[0].word.toUpperCase()}`
    );
    await user.click(isActive);
    await user.click(hasReminderOnload);
    await user.click(createButton);

    const notification = screen.getByRole("dialog", {
      name: "A word reminder has been created!",
    });
    expect(notification).toBeInTheDocument();
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockRegister).toHaveBeenCalledTimes(1);
    expect(mockRegister).toHaveBeenCalledWith();
    expect(mockCreateWordReminder).toHaveBeenCalledTimes(1);
    expect(mockCreateWordReminder).toHaveBeenCalledWith({
      userId: String(testUser.id),
      body: {
        reminder: "* * * * *",
        finish: new Date("2025-02-04"),
        user_words: [userWord1, userWord2],
        is_active: false,
        has_reminder_onload: false,
      },
    });
    expect(mockInvalidateQueries).toHaveBeenCalledTimes(1);
    expect(mockInvalidateQueries).toHaveBeenCalledWith({
      queryKey: ["wordReminders"],
    });
  });

  it("throws an error when the getWordList query fails", async () => {
    vi.spyOn(userWordService, "getUserWordList").mockImplementation(
      async () => {
        return { json: { message: new Error("Error Message.") }, status };
      }
    );
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });

    setup({ toggleModal: mockToggleModal, queryClient });

    const errorBoundary = await screen.findByTestId("error-boundary");
    expect(errorBoundary).toBeInTheDocument();
  });

  it("calls the functions to show a notification error", async () => {
    const message = "Bad Request.";
    const status = 400;
    vi.spyOn(userWordService, "getUserWordList").mockImplementation(
      async () => {
        return { json: { userWords: [userWord1, userWord2] }, status };
      }
    );
    const mockCreateWordReminder = vi
      .spyOn(wordReminderService, "createWordReminder")
      .mockImplementation(async () => {
        return Promise.reject({ json: { message }, status });
      });
    const mockRegister = vi.fn();
    vi.spyOn(
      useMobilePushNotificationHooks,
      "useMobilePushNotificationRegistration"
    ).mockReturnValue({ register: mockRegister });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const reminder = screen.getByLabelText("Reminder");
    const finish = screen.getByLabelText("Finish", { selector: "input" });
    const isActive = screen.getByLabelText("Is Active", { selector: "input" });
    const hasReminderOnload = screen.getByLabelText("Has Reminder Onload", {
      selector: "input",
    });
    const userWords = await screen.findByRole("combobox");
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(reminder, "* * * * *");
    await user.type(finish, "2025-02-04");
    await user.type(
      userWords,
      `${userWord1.details[0].word},${userWord2.details[0].word}`
    );
    await user.click(isActive);
    await user.click(hasReminderOnload);
    await user.click(createButton);

    const notification = screen.getByRole("dialog", {
      name: message,
    });
    expect(notification).toBeInTheDocument();
    expect(mockCreateWordReminder).toHaveBeenCalledTimes(1);
    expect(mockCreateWordReminder).toHaveBeenCalledWith({
      userId: String(testUser.id),
      body: {
        reminder: "* * * * *",
        finish: new Date("2025-02-04"),
        user_words: [userWord1, userWord2],
        is_active: false,
        has_reminder_onload: false,
      },
    });
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockRegister).not.toHaveBeenCalled();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
  });

  it("disables the create button when the mutation is loading", async () => {
    vi.spyOn(userWordService, "getUserWordList").mockImplementation(
      async () => {
        return { json: { userWords: [userWord1, userWord2] }, status };
      }
    );
    const mockCreateWordReminder = vi
      .spyOn(wordReminderService, "createWordReminder")
      .mockImplementation(async () => {
        const delay = 50;
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve({ json: { wordReminder }, status });
          }, delay);
        });
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });
    const reminder = screen.getByLabelText("Reminder");
    const finish = screen.getByLabelText("Finish", { selector: "input" });
    const userWords = await screen.findByRole("combobox");
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(reminder, "* * * * *");
    await user.type(finish, "2025-02-04");
    await user.type(userWords, `${userWord1.details[0].word}`);
    await user.click(createButton);

    expect(createButton).toBeDisabled();
    expect(mockCreateWordReminder).toHaveBeenCalledTimes(1);
    expect(mockCreateWordReminder).toHaveBeenCalledWith({
      userId: String(testUser.id),
      body: {
        reminder: "* * * * *",
        finish: new Date("2025-02-04"),
        user_words: [userWord1],
        is_active: true,
        has_reminder_onload: true,
      },
    });
    expect(mockToggleModal).not.toHaveBeenCalled();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
  });

  it("calls the functions to close a modal", async () => {
    const mockCreateWordReminder = vi
      .spyOn(wordReminderService, "createWordReminder")
      .mockImplementation(async () => {
        return { json: { wordReminder }, status };
      });
    const mockToggleModal = vi.fn();
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const { user } = setup({ toggleModal: mockToggleModal, queryClient });

    const closeButton = screen.getByRole("button", { name: "Close modal" });
    await user.click(closeButton);

    const notification = screen.queryByRole("dialog");
    expect(notification).not.toBeInTheDocument();
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
    expect(mockCreateWordReminder).not.toHaveBeenCalled();
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
  });

  describe("form validation", () => {
    it("does not allow user to submit without the reminder field", async () => {
      vi.spyOn(userWordService, "getUserWordList").mockImplementation(
        async () => {
          return {
            json: { userWords: [userWord1, userWord2] },
            status,
          };
        }
      );
      const mockCreateWordReminder = vi
        .spyOn(wordReminderService, "createWordReminder")
        .mockImplementation(async () => {
          return { json: { wordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient({
        defaultOptions: { queries: { retry: false } },
      });
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const reminder = screen.getByLabelText("Reminder");
      const finish = screen.getByLabelText("Finish", { selector: "input" });
      const userWords = await screen.findByRole("combobox");
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.clear(reminder);
      await user.type(finish, "2025-02-04");
      await user.type(userWords, `${userWord1.details[0].word}`);
      await user.click(createButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockCreateWordReminder).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });

    it("does not allow user to submit without the user_words field", async () => {
      vi.spyOn(userWordService, "getUserWordList").mockImplementation(
        async () => {
          return {
            json: { userWords: [userWord1, userWord2] },
            status,
          };
        }
      );
      const mockCreateWordReminder = vi
        .spyOn(wordReminderService, "createWordReminder")
        .mockImplementation(async () => {
          return { json: { wordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient({
        defaultOptions: { queries: { retry: false } },
      });
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const reminder = screen.getByLabelText("Reminder");
      const finish = screen.getByLabelText("Finish", { selector: "input" });
      const userWords = await screen.findByRole("combobox");
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.type(reminder, "* * * * *");
      await user.type(finish, "2025-02-04");
      await user.clear(userWords);
      await user.click(createButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockCreateWordReminder).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });

    it("does not allow user to submit without the finish field", async () => {
      vi.spyOn(userWordService, "getUserWordList").mockImplementation(
        async () => {
          return {
            json: { userWords: [userWord1, userWord2] },
            status,
          };
        }
      );
      const mockCreateWordReminder = vi
        .spyOn(wordReminderService, "createWordReminder")
        .mockImplementation(async () => {
          return { json: { wordReminder }, status };
        });
      const mockToggleModal = vi.fn();
      const queryClient = new QueryClient({
        defaultOptions: { queries: { retry: false } },
      });
      const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
      const { user } = setup({ toggleModal: mockToggleModal, queryClient });
      const reminder = screen.getByLabelText("Reminder");
      const finish = screen.getByLabelText("Finish", { selector: "input" });
      const userWords = await screen.findByRole("combobox");
      const createButton = screen.getByRole("button", { name: "Create" });

      await user.type(reminder, "* * * * *");
      await user.clear(finish);
      await user.type(userWords, `${userWord1.details[0].word}`);
      await user.click(createButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockToggleModal).not.toHaveBeenCalled();
      expect(mockCreateWordReminder).not.toHaveBeenCalled();
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });
  });
});
