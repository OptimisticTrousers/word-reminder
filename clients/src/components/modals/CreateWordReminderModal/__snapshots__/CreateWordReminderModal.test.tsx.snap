// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateWordReminderModal component > renders the form with all inputs 1`] = `[Function]`;
